### PR TITLE
Save scope initialization errors 

### PR DESCRIPTION
**Pull request by olivergrabinski**

### CODE REVIEW

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala [hunk 1]
```diff
@@ -33,15 +34,29 @@ final class ProjectsImpl private (
...
+                          case Left(e)  =>
+                            errorStore
+                              .save(init.entityType, resource.value.ref, e)
+                              .onError(e =>
+                                logger.error(e)(
+                                  s"Failed to save error for '${init.entityType}' initialization step on project '${resource.value.ref}'"
+                                )
+                              )
+                          case Right(_) => IO.unit
+                        }
```
Comment (shinyhappydan) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:55]: There will be something along the lines of `onError { case e: ScopeInitializationFailed => ???}` you could use

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala [hunk 2]
```diff
@@ -33,15 +34,29 @@ final class ProjectsImpl private (
...
+                        .onProjectCreation(resource.value, caller)
+                        .attemptNarrow[ScopeInitializationFailed]
+                        .flatMap {
+                          case Left(e)  =>
+                            errorStore
+                              .save(init.entityType, resource.value.ref, e)
+                              .onError(e =>
+                                logger.error(e)(
+                                  s"Failed to save error for '${init.entityType}' initialization step on project '${resource.value.ref}'"
+                                )
```
Comment (shinyhappydan) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:52]: The store should probably be doing this?

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    1,
+    entityType.value,
+    project.organization.value,
+    project.project.value,
+    scopeInitError.reason,
+    Instant.EPOCH
+  )
+
+  test("Inserting an error should succeed") {
+    assertStoreIsEmpty >>
```
Comment (shinyhappydan) [inline delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:41]: Would it ever not be empty?

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  case class ScopeInitErrorRow(
+      ordering: Int,
+      entityType: String,
+      org: String,
+      project: String,
+      message: String,
+      instant: Instant
+  )
+
+  implicit class ScopeInitErrorStoreOps(store: ScopeInitializationErrorStore) {
```
Comment (shinyhappydan) [inline delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:55]: I'm not sure what the implicit class is adding here, it's actually quite confusing that the store appears to have this method on it

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala [hunk 5]
```diff
@@ -265,5 +265,27 @@ class ProjectsImplSpec extends CatsEffectSpec with DoobieScalaTestFixture with C
...
+      }
+
+      val inits    = Set(failingScopeInit, successfulScopeInit)
+      val projects = ProjectsImpl(fetchOrg, validateDeletion, inits, defaultApiMappings, config, xas, clock)
+
+      val createProject                    = projects.create(project, payload)
+      val assertSuccessfulInitStepExecuted =
+        successfulScopeInit.createdProjects.get.map(p => p shouldEqual Set(project))
+
+      createProject >> assertSuccessfulInitStepExecuted
```
Comment (shinyhappydan) [inline delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala:288]: Should it assert that a log was recorded?

**Review by shinyhappydan — APPROVED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala (hunk 1 above)
Comment (olivergrabinski) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:55]: You're right, I already had the same issue at some point; I keep forgetting about `recover`

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala (hunk 2 above)
Comment (olivergrabinski) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:52]: Yes I was going to move that later

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala (hunk 5 above)
Comment (olivergrabinski) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala:288]: Yes there should be a check for that. The reason I did not add it is because I didn't want to have to deal with the errorStore directly. In my mind the next step is the ability to check the health/errors for a given project which can then be used to do that check

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala (hunk 3 above)
Comment (olivergrabinski) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:41]: If I were to add another test now, it would start with a non empty store. Should it be more like a `resetStore` method, here? Not sure

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala (hunk 4 above)
Comment (olivergrabinski) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:55]: My thinking here was that I didn't want to add methods to the actual implementation that were only used in tests. However I agree it's not very successful, and it should be gone soon, if not in this PR then in the next one

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala (hunk 3 above)
Comment (shinyhappydan) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:41]: I think the tests having a clean slate isn't really something you need to check at the beginning of the test, it's more for a beforeAll

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala (hunk 4 above)
Comment (shinyhappydan) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:55]: I see what you mean, but I would suggest that a normal method would be clearer than the implicit class

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala (hunk 5 above)
Comment (shinyhappydan) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImplSpec.scala:288]: I think it's legit to deal with the error store (mocking it probably makes the most sense for this test), but your suggested next step could work too

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala (hunk 4 above)
Comment (imsdu) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ScopeInitializationErrorStoreSuite.scala:55]: As those information should be part of the new endpoint allowing to check that projects are correctly initialized.

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala [hunk 6]
```diff
@@ -33,15 +34,20 @@ final class ProjectsImpl private (
...
+  )(implicit caller: Subject): IO[ProjectResource] =
     for {
       resource <- eval(CreateProject(ref, fields, caller)).span("createProject")
       _        <- scopeInitializations
-                    .parUnorderedTraverse(_.onProjectCreation(resource.value, caller))
+                    .parUnorderedTraverse { init =>
+                      init
+                        .onProjectCreation(resource.value, caller)
+                        .recoverWith { case e: ScopeInitializationFailed =>
+                          errorStore.save(init.entityType, resource.value.ref, e)
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:45]: The error should still be returned to the user

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala [hunk 7]
```diff
@@ -33,15 +34,20 @@ final class ProjectsImpl private (
   override def create(
       ref: ProjectRef,
       fields: ProjectFields
-  )(implicit caller: Subject): IO[ProjectResource] = {
+  )(implicit caller: Subject): IO[ProjectResource] =
     for {
       resource <- eval(CreateProject(ref, fields, caller)).span("createProject")
       _        <- scopeInitializations
-                    .parUnorderedTraverse(_.onProjectCreation(resource.value, caller))
+                    .parUnorderedTraverse { init =>
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/projects/ProjectsImpl.scala:41]: A bit like `IndexingAction`, it could be interesting to refactor this by introducing some kind of aggregate there which would take care of the init steps and saving errors
